### PR TITLE
Define MADV_FREE on our own

### PR DESCRIPTION
**Pull request by interwq**
On x86 Linux, we define our own MADV_FREE if madvise(2) is available, but no
MADV_FREE is detected.  This allows the feature to be built in and enabled with
runtime detection.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -561,7 +561,7 @@ case "${host}" in
 	dnl syscall(2) and secure_getenv(3) are exposed by _GNU_SOURCE.
 	JE_APPEND_VS(CPPFLAGS, -D_GNU_SOURCE)
 	abi="elf"
-	AC_DEFINE([JEMALLOC_PURGE_MADVISE_DONTNEED_ZEROS])
+	AC_DEFINE([JEMALLOC_PURGE_MADVISE_DONTNEED_ZEROS], [ ])
```
Comment (davidtgoldblatt) [inline configure.ac]: What's the difference between these?

Comment (interwq) [reply configure.ac]: The default adds a value 1 to the macro (i.e. it does #define JEMALLOC_PURGE_MADVISE_DONTNEED_ZEROS 1) , which we don't use.

File: include/jemalloc/internal/jemalloc_preamble.h.in [hunk 2]
```diff
@@ -47,6 +47,10 @@
 #endif
 #include "jemalloc/internal/hooks.h"
 
+#ifdef JEMALLOC_DEFINE_MADVISE_FREE
+#  define MADV_FREE 8
```
Comment (davidtgoldblatt) [inline include/jemalloc/internal/jemalloc_preamble.h.in]: We should probably prefix this somehow, to avoid naming conflicts with the builtin, no?

Comment (interwq) [reply include/jemalloc/internal/jemalloc_preamble.h.in]: JEMALLOC_DEFINE_MADVISE_FREE will only present when MADV_FREE is not found on the system. I assume it doesn't matter in this case?

Comment (davidtgoldblatt) [reply include/jemalloc/internal/jemalloc_preamble.h.in]: Eh, probably not a big deal.

Comment (interwq) [reply include/jemalloc/internal/jemalloc_preamble.h.in]: Well, I'll see if we can make it clean with a prefix.

File: src/pages.c [hunk 3]
```diff
@@ -269,7 +269,13 @@ pages_purge_lazy(void *addr, size_t size) {
...
 	return false;
 #elif defined(JEMALLOC_PURGE_MADVISE_FREE)
-	return (madvise(addr, size, MADV_FREE) != 0);
+	return (madvise(addr, size,
+#  ifdef MADV_FREE
+	    MADV_FREE
+#  else
+	    JEMALLOC_MADV_FREE
+#  endif
+	    ) != 0);
```
Comment (interwq) [inline src/pages.c]: @davidtgoldblatt: this should make it safer.

**Pull request merged**